### ISSUE TITLE
How do you extract the weights used in the model?

### ISSUE THREAD

**Issue by thoppe** [google/prettytensor#6]
### Request for documentation

I've trained a 3-layer fully connected layer as a test and it works perfectly. I'd like to see what the weights used in each layer are but I'm not sure how access them through prettytensor. In standard tensorflow the variables are explicit, but here they hidden through the "pretty" interface. 

An example in the docs to shows how to access the weights would be very useful for others. We can use the example you have in the docs:

    result = (pretty_tensor.wrap(input_data, m)
          .flatten()
          .fully_connected(200, activation_fn=tf.nn.relu)
          .fully_connected(10, activation_fn=None)
          .softmax(labels, name=softmax_name))

as an example to extract the weights.

**Comment by eiderman** [comment]
TensorFlow provides graph collections that group the variables.  To access the variables that were trained you would call `tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)` or to get all variables (including some for statistics) use `tf.get_collection(tf.GraphKeys.VARIABLES)` or its shorthand `tf.all_variables()`

**Comment by thoppe** [comment]
Thanks, this answers my question completely.

**Comment by PatWie** [comment]
This is still an *open* issue.Having

```
ans = (pt.wrap(input).flatten().fully_connected(10,name='layer1').fully_connected(10,name='layer2')).tensor
sess.run(ans,feed_dict={input: myinput}
```

works as expected. But 
```
print sess.run(tf.get_variable('layer1') )
```

does not. Running `tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)` would give all variables without names. Or do we have to scope each layer?.

**Comment by eiderman** [comment]
fully_connected creates 2 variables: weights and bias. As such each is scoped to 'layer1' and has their own names.

You should be able to get them with `tf.get_variable('layer1/weights')` and `'layer1/bias'`, but you need to make sure that your variable scope is in reuse mode:

```
with tf.variable_scope('layer1', reuse=True):
  tf.get_variable('weights')
```

Other ways to get a variable:

1.  If you break the chain, you can call `layer_parameters['weights']` on the Pretty Tensor object.
2. To get all variables associated with a layer, you can use: `tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, 'layer1')`
3. To get all weights (and not bias):
```
def get_weights():
  return [v for v in tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES) if v.name.endswith('weights:0')]
```

Note the ':0' which is added because the variable is actually the first result of a variable_op.

**Comment by devmessias** [comment]
How do you extract the weights used in the model in EACH training step(each update of weight matrix)? I need save these matrices in  disk,

**Comment by shirishr** [comment]
See if this helps:


http://cv-tricks.com/tensorflow-tutorial/save-restore-tensorflow-models-quick-complete-tutorial/

[http://cv-tricks.com/wp-content/uploads/2017/03/tensorflow-logo.png]<http://cv-tricks.com/tensorflow-tutorial/save-restore-tensorflow-models-quick-complete-tutorial/>

save and restore Tensorflow models - CV-Tricks.com<http://cv-tricks.com/tensorflow-tutorial/save-restore-tensorflow-models-quick-complete-tutorial/>
cv-tricks.com
In this quick Tensorflow tutorial, you shall learn what's a Tensorflow model and how to save and restore Tensorflow models for fine-tuning and building on top of them.





Bye for now,
Shirish


________________________________

**Comment by devmessias** [comment]
Thanks, @shirishr. But that's not it. Every time the batch ends, I need extract the weight matrix ...

**Comment by idoh** [comment]
@devmessias did you find a solution?

**Comment by alexbui91** [comment]
@devmessias Did you find the solution?

**Comment by bshao001** [comment]
A tutorial explaining how to analyze, deploy and handle TensorFlow and Keras models is available here: https://github.com/bshao001/TF-Model-Deploy-Tutorial. Check it out and see if that helps.

**Comment by wspstorm** [comment]
Hi, do you have an example of c++ code to extract the weights?

**Comment by rocksyne** [comment]
This may be old, but just keeping it here for the future ^..^

**Python Code**

```
tvars = tf.trainable_variables()
tvars_vals = sess.run(tvars)

for var, val in zip(tvars, tvars_vals):
    print(var.name, val)  # Prints the name of the variable alongside its value.
```

**Comment by joriatyBen** [comment]
thanks rcksyne, this works, repectively there are weights, but for me it is not sure that they are trained already. I used that weights and feed another indedpendent pure python model and the result is really bad.
